sled-agent/src/instance.rs

```diff
@@ -477,74 +533,147 @@ impl InstanceRunner {
+                                        snapshot_id
+                                    ).await.map_err(|e| e.into())
+                                )
+                                .map_err(|_| Error::FailedSendClientClosed)
+                            },
+                            AddExternalIp { ip, tx } => {
+                                tx.send(self.add_external_ip(&ip).await.map_err(|e| e.into()))
+                                .map_err(|_| Error::FailedSendClientClosed)
+                            },
+                            DeleteExternalIp { ip, tx } => {
+                                tx.send(self.delete_external_ip(&ip).await.map_err(|e| e.into()))
+                                .map_err(|_| Error::FailedSendClientClosed)
+                            },
+                            RefreshExternalIps { tx } => {
+                                tx.send(self.refresh_external_ips().map_err(|e| e.into()))
+                                .map_err(|_| Error::FailedSendClientClosed)
+                            }
+                        }
+                    };
+                    tokio::select! {
+                        biased;
+
+                        request = terminate_rx.recv() => {
+                            match request {
+                                Some(TerminateRequest { tx, mark_failed }) => {
+                                    info!(
+                                        self.log,
+                                        "Received request to terminate instance \
+                                         while waiting on an ongoing request";
+                                        "request" => request_variant,
+                                    );
+                                    let result = tx.send(Ok(VmmUnregisterResponse {
+                                        updated_runtime: Some(self.terminate(mark_failed).await)
+                                    }))
+                                    .map_err(|_| Error::FailedSendClientClosed);
+                                if let Err(err) = result {
+                                        warn!(
+                                            self.log,
+                                            "Error handling request to terminate instance";
+                                            "err" => ?err,
+                                        );
+                                    }
+                                    break;
+                                },
+                                None => {
+                                    warn!(
+                                        self.log,
+                                        "Instance termination request channel closed; \
+                                        shutting down";
+                                    );
+                                    self.terminate(false).await;
+                                    break;
+                                },
```

Comment:
This seems like it could maybe be deduplicated with the other code on line ~539 which handles the `terminate_rx` the same way?

Reply:
Yeah, some of it probably could be de-duplicated. The reason it's not is because I wanted the case where the termination request was received while in the middle of another instance op to be able to log what was interrupted by the termination request, which seemed potentially useful for debugging "why did this instance operation abruptly vanish?". I'll see if there's a way to preserve that while making it less repetitive.

Reply:
I also realized that rustfmt has basically given up on  formatting this because it's inside a macro, which is also an argument in favor of pulling it out...